### PR TITLE
Use WeakRef on tree hook

### PR DESCRIPTION
**Pull request by dapplion**
Investigating a running Lodestar node, we found that there were many more merkle nodes in memory than we expected.
By inspecting a heap snapshot, we were able to track down the source of much of the discrepancy.

Our mutable `Tree` wrapper optionally contains a `Hook` with a reference to a parent `Tree`.
As a result, if subtrees are kept, accidentally or otherwise, the parent trees will never be garbage collected.

This is a proposal to wrap the Tree hook with a [`WeakRef`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef) so it doesn't prevent garbage collection of the parent.

TODO:
- [x] Feature detection, this is extra and should work when WeakRef not available
- [x] Add tests

### CODE REVIEW

File: src/tree.ts [hunk 1]
```diff
@@ -10,22 +10,47 @@ const ERR_INVALID_TREE = "Invalid tree operation";
 const ERR_PARAM_LT_ZERO = "Param must be >= 0"
 const ERR_COUNT_GT_DEPTH = "Count extends beyond depth limit"
 
+enum HooKTypeCode {
```
Comment (wemeetagain) [inline src/tree.ts:13]:
```suggestion
enum HookTypeCode {
```

**Comment by dapplion** [comment]
Test what happens if a dependant requires this and does not declare `lib: ["esnext"]`

File: test/tree.test.ts [hunk 2]
```diff
@@ -36,3 +36,27 @@ describe("fixed-depth tree iteration", () => {
     }
   });
 });
+
+describe("subtree mutation", () => {
+  it.only("changing a subtree should change the parent root", () => {
```
Comment (tuyennhv) [inline test/tree.test.ts:41]: no `only`? :D

Comment (tuyennhv) [reply test/tree.test.ts:41]: having unit tests in this repo is so nice btw

**Comment by dapplion** [comment]
I've linked this branch to ssz, link to lodestar and run with node v14 ensuring WeakRef is available

Screenshot from 2021-03-25 12-44-40

**Review by wemeetagain — APPROVED** [review]
:tada:

**Pull request merged**